### PR TITLE
feat: [Custom dimension] Binning with fixed number

### PR DESCRIPTION
**Pull request by rephus**
Closes: #5198

### Description

Allow users to define a dimension that is going to return a fixed amount of bins. 

Design mockups:
https://www.figma.com/file/JLRWyriHFvNnGu6F3klL8o/Untitled?node-id=0-1

- [x] only numeric dimensions can be 'binned'
- [x] if you have a numeric dimension, you can open the three-dot menu and see the option to `create custom dimension` (within this, there is the option to select `bin`)
- [x] There is a modal that pops up to create a binned custom dimension
- [x] You must name your custom dimension.
  - [x] By default, this is just `[name of dimension] bins` (e.g. creating a bin on a dimension `age` would be named `age bins` by default)
- [x] There are no `Bin Types` section (yet) - the only option is fixed number of bins for this ticket.
- [x] There is a number picker (like in filters) where you select the number of bins that you want. 
- [x] This number must be an integer.
- [ ] There is copy below the number of bins which tells you what the bin ranges will be for your data 
- [ ] There is copy that warns the user that these bins might change if their underlying data changes. 
- [ ] Once you've saved the custom dimension, it appears on the bottom of the sidebar under `custom dimensions`, above the `metrics` section, but below the `dimensions` section 
- [ ] The custom dimension is selected by default and added to your table. It behaves like a custom metric (for deletion and persistence)

Extra acceptance criteria: 

- [x] Make sure it works for charts
- [x] Sort bins
- [x] Make sure you can select this custom dimension on a chart
- [x] Chart tooltips 
- [x] Pie charts
- [x] Grouped charts
- [x] Pivot tables https://github.com/lightdash/lightdash/issues/7563
- [x] table charts
- [x] You can delete a custom dimension from explore 
- [x] It works with filters  https://github.com/lightdash/lightdash/issues/7564
- [x] It works with sorting
- [x] Build unit tests for bulidquery
- [x] Build e2e tests for custom dimension https://github.com/lightdash/lightdash/issues/7580
- [x] Disable `non-number` inputs in bins
- [ ] Work on `preview results` on the FE
- [ ] Rework `toggleActiveField`
- [x] Test underlying dasta - created a ticket
- [x] Wrap this in feature flag
- [x] Replace getCustomDimensionId with unique label toa void conflicts with other custom dimensions
- [x] Test multiple custom dimensions from the same dimension
- [x] Reference lines - created a ticket
- [x] CTE compatible with table calculations
- [x] persistence (migration + get + save) 
- [x] 1 bucket only edge case
- [x] multiple joins 
- [x] Test on dashboards 

Alternative SQL implementation https://github.com/lightdash/lightdash/pull/7545

Technical details:
1. add custom dimension to `MetricQuery` type. We should basically follow what we did for the custom metric ( aka additional metric ) 
2. if a binning custom dimension is selected, the query builder should:
  - add the "original" dimension in the main query
  - add the necessary sql in the second query to create the bins
  
 https://stackoverflow.com/questions/15464574/use-google-bigquery-to-build-histogram-graph
 

3. if we are sorting by the binning custom dimension, the query builder should:
  - add a custom order by 
```
// example
ORDER BY CASE `age`
         WHEN '< 0' THEN 1
         WHEN '0 to 100' THEN 2
         WHEN '100+' THEN 3
         ELSE 4
         END,
         `age`  
```
4. in the chart visualization, if the binning custom dimension is used, we need to add custom sorting to the series. Otherwise, it will be sorted alphabetically. 

Captura de ecrã 2023-04-25, às 16 21 42

### What problem does this solve?

As a user, I want to be able to create histograms so that I can see how my data is distributed across a specific number of groups.
### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/common/src/index.ts [hunk 1]
```diff
@@ -694,7 +696,7 @@ export const getAxisName = ({
     axisIndex: number;
     axisName?: string;
     series?: Series[];
-    items: Array<Field | TableCalculation>;
+    items: Array<Field | TableCalculation | CustomDimension>;
```
Comment (rephus) [inline packages/common/src/index.ts:699]: Potential things to refactor: 

We should use unique types (from common) and avoid joining types, this way we don't need to update methods like this: 

eg: AxisItems = Field | TableCalculation | CustomDimension

File: packages/common/src/types/field.ts [hunk 2]
```diff
@@ -103,7 +119,12 @@ export type TableCalculation = {
 };
 
 export const isTableCalculation = (item: Item): item is TableCalculation =>
-    item ? !!item.sql && !('type' in item) && !('tableName' in item) : false;
+    item
+        ? !('binType' in item) &&
```
Comment (rephus) [inline packages/common/src/types/field.ts:123]: Potential things to refactor: 

This type checking is not very reliable. I had to add `!('binType' in item)` after I added CustomDimension to Item becuase it was getting confused

**Comment by rephus** [comment]
You can have multiple custom dimensions (aka multiple CTEs) 
image

**Comment by rephus** [comment]
Good news about the SQL builder, even though I'm introducing many changes and new CTEs, I kept the SQL withotu custom dimensions the same , so old tests are still passing 

image

File: packages/backend/src/queryBuilder.ts [hunk 3]
```diff
@@ -128,6 +131,105 @@ const getJoinType = (type: DbtModelJoinType = 'left') => {
     }
 };
 
+export const getCustomDimensionSql = ({
```
Comment (rephus) [inline packages/backend/src/queryBuilder.ts:134]: Potential things to refactor:

I extracted all the logic for custom dimensions on a method, this way is easier to generate the sql, we should probably do the same for table calculations, etc.

File: packages/backend/src/queryBuilder.ts [hunk 4]
```diff
@@ -393,7 +509,11 @@ export const buildQuery = ({
             sqlGroupBy,
         ].join('\n');
         const cteName = 'metrics';
-        const cte = `WITH ${cteName} AS (\n${cteSql}\n)`;
+        const ctes = [
```
Comment (rephus) [inline packages/backend/src/queryBuilder.ts:512]: Potential things to refactor:

This is a special edgecase that forces us to wrap the whole query into a CTE so we can use table calculations.

This makes me think we should probably have a better way to join CTEs in the future.

**Comment by rephus** [comment]
Restarted db on render to apply migration

**Comment by rephus** [comment]
Tested with grouping + save + dashboards 

image

File: packages/backend/src/@types/knex-tables.d.ts [hunk 5]
```diff
@@ -171,7 +173,7 @@ declare module 'knex/types/tables' {
         [SavedChartVersionsTableName]: SavedChartVersionsTable;
         [SavedChartVersionFieldsTableName]: SavedChartVersionFieldsTable;
         [SavedChartVersionSortsTableName]: SavedChartVersionSortsTable;
-        [SavedQueryTableCalculationTableName]: SavedQueryTableCalculationTable;
+        [SavedChartTableCalculationTableName]: SavedQueryTableCalculationTable;
```
Comment (rephus) [inline packages/backend/src/@types/knex-tables.d.ts:176]: This was a typo on table calculations, spotted here and fixed.

File: packages/backend/src/@types/knex-tables.d.ts [hunk 6]
```diff
@@ -171,7 +173,7 @@ declare module 'knex/types/tables' {
         [SavedChartVersionsTableName]: SavedChartVersionsTable;
         [SavedChartVersionFieldsTableName]: SavedChartVersionFieldsTable;
         [SavedChartVersionSortsTableName]: SavedChartVersionSortsTable;
-        [SavedQueryTableCalculationTableName]: SavedQueryTableCalculationTable;
+        [SavedChartTableCalculationTableName]: SavedChartTableCalculationTable;
```
Comment (rephus) [inline packages/backend/src/@types/knex-tables.d.ts:176]: This was a typo on table calculations, spotted here and fixed

**Comment by joaoviana** [comment]
> Looks really good - a few things noted:
> 
> 1. I shouldn't see this "edit calculation button" that then opens the table calc modal:

on `1`, I have a PR that will be ready for review soon: https://github.com/lightdash/lightdash/pull/7590 screen-recording incoming

**Comment by hamzahc1** [comment]
One more thing @rephus - pie charts aren't working for me. Whenever I select Pie chart, the chart section goes blank.

**Comment by rephus** [comment]
> 2\. Using "to" makes sense in the modal when creating a dimension, but on the chart and in results table etc we should use a hyphen (e.g. 0-10, 10-20 not 0 to 10, 10 to 20)

The `to` was part of a screenshot on the spec. I don't mind switching, it should be a really simple change. 

> I couldn't test this in the PR - but do the bin widths automatically update if the min/max values of the dimensions change?

yes becuase the range is calculated in realtime in  the SQL, we don't store the current values on the custom dimension

**Comment by hamzahc1** [comment]
> The `to` was part of a screenshot on the spec. I don't mind switching, it should be a really simple change.
> 
> >

Yep let's change it for hyphens! 

> mmm, it seems ok to me . What can I do to replicate your issue ?

Just tried it again, seems to work! May have been a one-off problem I had in the testing env 🤷

File: packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+    const { setFieldValue } = form;
+
+    useEffect(() => {
+        if (isEditing && isCustomDimension(item)) {
+            setFieldValue('customDimensionLabel', item.name);
+            setFieldValue('binType', item.binType);
+            setFieldValue(
+                'binConfig.fixedNumber.binNumber',
+                item.binNumber ? +item.binNumber : 0,
```
Comment (rephus) [inline packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx:78]: should the default be 2 ?

File: packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+                            </Tooltip>
+                        </Group>
+                    </Radio.Group>
+
+                    {form.values.binType === BinType.FIXED_NUMBER && (
+                        <NumberInput
+                            w={100}
+                            label="Bin number"
+                            required
+                            min={2}
```
Comment (rephus) [inline packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx:187]: should we set a max ?

File: packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNode.tsx [hunk 9]
```diff
@@ -51,10 +57,19 @@ const TreeSingleNode: FC<Props> = ({ node }) => {
...
     const isFiltered = isField(item) && isFilteredField(item);
 
-    const label = timeIntervalLabel || item.label || item.name;
-
+    const label = isField(item) ? timeIntervalLabel || item.label : item.name;
+    const description = isField(item) ? item.description : undefined;
     const bgColor = getItemBgColor(item);
 
+    const getFieldIconColor = (field: Item | AdditionalMetric) => {
+        if (isCustomDimension(field) || isDimension(field)) return 'blue.9';
```
Comment (rephus) [inline packages/frontend/src/components/Explorer/ExploreTree/TableTree/Tree/TreeSingleNode.tsx:65]: Potential things to refactor, 

Instead of doing all these `if is type` methods, we could have a function (getFieldType) that return sa string, that we can use on a `switch`

File: packages/frontend/src/components/VisualizationConfigs/PieChartConfig/PieChartLayoutConfig.tsx [hunk 10]
```diff
@@ -23,23 +24,25 @@ const PieChartLayoutConfig: React.FC = () => {
         },
     } = useVisualizationContext();
 
+    const allDimensions = [...dimensions, ...customDimensions];
```
Comment (rephus) [inline packages/frontend/src/components/VisualizationConfigs/PieChartConfig/PieChartLayoutConfig.tsx:27]: @joaoviana we already do this on pie charts :eyes:

**Review by rephus — COMMENTED** [review]
@joaoviana your changes look good to me :+1:

File: packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx (hunk 8 above)
Comment (rephus) [reply packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx:187]: image
 :see_no_evil: it doesn't error

Comment (rephus) [inline packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx:187]: I thought you said we should allow `1` , (the backend supports it now)

**Review by joaoviana — APPROVED** [review]

File: packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx (hunk 7 above)
Comment (joaoviana) [reply packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx:78]: let's go with `1` since the backend allows that now?

File: packages/frontend/src/hooks/tableVisualization/getDataAndColumns.tsx [hunk 11]
```diff
@@ -85,8 +86,7 @@ const getDataAndColumns = ({
                                 <TableHeaderBoldLabel>
                                     {item === undefined
                                         ? 'Undefined'
-                                        : item.displayName ||
-                                          friendlyName(item.name)}
+                                        : item.name || friendlyName(item.name)}
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/frontend/src/hooks/tableVisualization/getDataAndColumns.tsx:89]: Yes, good catch

File: packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx (hunk 8 above)
Comment (joaoviana) [reply packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx:187]: Let's leave it for now and then we can find through user feedback 😄  https://github.com/lightdash/lightdash/issues/7600

File: packages/frontend/src/hooks/cartesianChartConfig/useCartesianChartConfig.ts [hunk 12]
```diff
@@ -364,17 +365,28 @@ const useCartesianChartConfig = ({
...
         return [
-            [...sortedDimensions, ...metrics, ...tableCalculations],
-            sortedDimensions,
+            [
+                ...sortedDimensions,
+                ...metrics,
+                ...tableCalculations,
+                ...customDimensions,
+            ],
+            [...sortedDimensions, ...customDimensions],
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/frontend/src/hooks/cartesianChartConfig/useCartesianChartConfig.ts:386]: Yes, this is used to return `customdimensions`on the list of `availableDimensions` which is going to be used to  `Set fallout layout values`

File: packages/backend/src/database/migrations/20231020102654_custom_dimensions.ts [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+                    'saved_queries_version_custom_dimension_id',
+                    'integer PRIMARY KEY GENERATED ALWAYS AS IDENTITY',
+                );
+                tableBuilder.text('id').notNullable();
+                tableBuilder.text('name').notNullable();
+                tableBuilder.text('dimension_id').notNullable();
+                tableBuilder.text('table').notNullable();
+                tableBuilder.integer('order').notNullable();
+                tableBuilder.text('bin_type').notNullable();
+                tableBuilder.text('bin_number').nullable();
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/database/migrations/20231020102654_custom_dimensions.ts:20]: yes

File: packages/frontend/src/components/VisualizationConfigs/PieChartConfig/PieChartLayoutConfig.tsx (hunk 10 above)
Comment (joaoviana) [reply packages/frontend/src/components/VisualizationConfigs/PieChartConfig/PieChartLayoutConfig.tsx:27]: So this would help on the toggling... correct? 🤔

File: packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx (hunk 8 above)
Comment (joaoviana) [reply packages/frontend/src/components/Explorer/CustomDimensionModal/index.tsx:187]: Amazing that it supports it now! changing to 2!

File: packages/backend/src/queryBuilder.test.ts [hunk 14]
```diff
@@ -390,3 +392,113 @@ describe('assertValidDimensionRequiredAttribute', () => {
         expect(result).toBeUndefined();
     });
 });
+
+describe('with custom dimensions', () => {
+    it('getCustomDimensionSql with empty custom dimension', async () => {
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/queryBuilder.test.ts:397]: you are right, I copied from the tests above

File: packages/frontend/src/hooks/cartesianChartConfig/useCartesianChartConfig.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (joaoviana) [reply packages/frontend/src/hooks/cartesianChartConfig/useCartesianChartConfig.ts:386]: Thanks for explaining!
